### PR TITLE
Checkpoint migration for loop's global step

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

WIP

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/tests_pytorch/utilities/test_upgrade_checkpoint.py [hunk 1]
```diff
@@ -1,49 +0,0 @@
...
-            {"epoch": 1, "global_step": 23, "checkpoint_callback_best_model_path": "path"},
-            {"epoch": 1, "global_step": 23, "callbacks": {ModelCheckpoint: {"best_model_path": "path"}}},
-        ),
-        (
-            {"epoch": 1, "global_step": 23, "early_stop_callback_wait": 2, "early_stop_callback_patience": 4},
-            {"epoch": 1, "global_step": 23, "callbacks": {EarlyStopping: {"wait_count": 2, "patience": 4}}},
-        ),
-    ],
-)
-def test_upgrade_checkpoint(tmpdir, old_checkpoint, new_checkpoint):
```
Comment (awaelchli) [inline tests/tests_pytorch/utilities/test_upgrade_checkpoint.py:43]: Moved this test to test_migration.py where it belongs after #15237

File: tests/tests_pytorch/utilities/migration/test_migration.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+from pytorch_lightning.utilities.migration import migrate_checkpoint
+from pytorch_lightning.utilities.migration.utils import _get_version, _set_legacy_version, _set_version
+
+
+@pytest.mark.parametrize(
+    "old_checkpoint, new_checkpoint",
+    [
+        (
+            {"epoch": 1, "global_step": 23, "checkpoint_callback_best": 0.34},
+            {"epoch": 1, "global_step": 23, "callbacks": {ModelCheckpoint: {"best_model_score": 0.34}}, "loops": ANY},
```
Comment (awaelchli) [inline tests/tests_pytorch/utilities/migration/test_migration.py:32]: This test was moved. only thing that was added to the "expected" state dict was the loops key.

File: tests/tests_pytorch/models/test_restore.py [hunk 3]
```diff
@@ -254,29 +254,6 @@ def on_train_start(self) -> None:
     assert trainer.fit_loop.epoch_loop._batches_that_stepped == max_epochs * train_batches
 
 
-@pytest.mark.parametrize("model_class", [BoringModel, ManualOptimBoringModel])
-def test_logging_step_loaded_correctly_pre_1_6_5(tmpdir, model_class):
```
Comment (awaelchli) [inline tests/tests_pytorch/models/test_restore.py:258]: this test moved to test_migration.py

File: tests/tests_pytorch/utilities/migration/test_migration.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+def test_migrate_loop_current_epoch_to_progress_tracking():
+    old_checkpoint = {"global_step": 15, "epoch": 2}
+    _set_version(old_checkpoint, "1.5.9")  # pretend a checkpoint prior to 1.6.0
+    updated_checkpoint, _ = migrate_checkpoint(old_checkpoint)
+    assert updated_checkpoint["loops"]["fit_loop"]["epoch_progress"]["current"]["completed"] == 2
+
+
+@pytest.mark.parametrize("model_class", [BoringModel, ManualOptimBoringModel])
+def test_migrate_loop_batches_that_stepped(tmpdir, model_class):
```
Comment (awaelchli) [inline tests/tests_pytorch/utilities/migration/test_migration.py:90]: only change to this test is the name and inserted `_set_version(ckpt, "1.6.4")` plus adjusted comment

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/utilities/migration/migration.py [hunk 5]
```diff
@@ -67,3 +69,94 @@ def _migrate_model_checkpoint_early_stopping(checkpoint: _CHECKPOINT) -> _CHECKP
...
+    Version: 1.6.5
+    Commit: c67b075
+    PR: #13645
+    """
+    global_step = checkpoint["global_step"]
+    checkpoint["loops"]["fit_loop"]["epoch_loop.state_dict"].setdefault("_batches_that_stepped", global_step)
+    return checkpoint
+
+
+_FIT_LOOP_INITIAL_STATE_1_6_0 = {
```
Comment (carmocca) [inline src/pytorch_lightning/utilities/migration/migration.py:121]: I would suggest just adding the fields that we will migrate. Loading should still work with a partial dictionary

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/utilities/migration/migration.py (hunk 5 above)
Comment (awaelchli) [reply src/pytorch_lightning/utilities/migration/migration.py:121]: I am not so confident about that. I tried but this makes it a bit more complicated. The current way IMO is easier to understand and it is better documented what the full state structure was in 1.6.0.
